WIP: hyperkernel mods

cjams opened the pull request.

Commit:
Add xAPIC write handler

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
WIP: insn decoder

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Add x2apic handlers

- Add default x2apic read and write handlers
- Add specific handlers for ICR, LVT timer, and TSC-deadline writes

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Use return_and_continue for SCHEDOP_yield

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Fix FADT checksum

- Fix FADT checksum by actually doing the checksum
- Passthrough PIC and ACPI SCI ports (for now)

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Move callback via to evtchn_op

When the guest uses HVMOP_set_param to set the callback via,
Xen calls hvm_assert_evtchn_irq, which ultimately results
in a vcpu_kick/interrupt injection.

We don't have a scheduler, so for now our vcpu_kick and
injection are one in the same. We queue an interrupt
at the callback vector, after setting the
evtchn_upcall_pending in the shared vcpu_info.

The evtchn_pending_sel field is set before the callback is
set_param'd in evtchn_op::setup_ports

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Add back xAPIC handlers

At this point, the guest is done with the intialization of its xAPIC
registers. No reads occur, only writes.

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Add cpuid and msr handlers

cpuid
- leaf 0x2: Pass-through cpuid cache info
- leaf 0x4: Set number of cpus to 0 in eax
- leaf 0xA: Disable performance monitoring

msr
- 0x140: Pass-through monitor/mwait
- 0x0CE: Pass-through MSR_PLATFORM_INFO. Note this is essentially
  already pass-through'd via the pvclock calculations.
  Specifically, the TSC frequency used in those calculations
  are derived from the max non-turbo ratio in this MSR

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Add debug dissassembly code

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Pass-through ports for PIT-based TSC calibration

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Add driver build files to .gitignore

Signed-off-by: Connor Davis <davisc@ainfosec.com>

rianquinn commented on bfvmm/include/hve/arch/intel_x64/lapic.h:
```diff
@@ -120,15 +120,116 @@ class EXPORT_HYPERKERNEL_HVE lapic
+    /// @note the index is a dword offset, not a byte offset
+    ///
+    void write(uint32_t idx, uint32_t val)
+    { m_lapic_view[idx] = val; }
```
Now that you have the x2APIC working again, is there a plan to delete all of the memory stuff? I would like to keep this to the bare minimum for now

cjams replied:
x2 doesn't work with PVH right now. specifically we would have to disable hvm_pvclock and hvm_pirqs

rianquinn replied:
☹️

rianquinn commented on bfvmm/include/hve/arch/intel_x64/vcpu.h:
```diff
@@ -258,6 +258,20 @@ class vcpu : public eapis::intel_x64::vcpu
+    /// @param indx the dword offset to read from
+    /// @return the 32-bit value of the register
+    ///
+    VIRTUAL uint32_t lapic_read(uint32_t indx) const;
```
Not a big fan of these. Why would someone using the vCPU need raw access to APIC registers? I'd prefer that if you need to know something that is in the vCPU that we provide an API to get it which removes the train wreck.

cjams replied:
When APIC accesses are being handled from an exit, you are working with an indx. If we had explicity APIs for each register, we would have giant switch statements in the exit handlers. That being said, I agree that things external to the VCPU shouldn't need this; this is more for cllients within the vcpu i.e. xen_op_handler. So we could use a different pattern like a friend or something

rianquinn replied:
Agreed. The APIC class should be self-contained, included decoding instructions, etc... and the ept_violation handler should be in the APIC class itself. The APIC class should then expose whatever APIs the other classes need like setting the TSC mode, etc... 

The Xen-ops class just handles the resources for now, and the APIC is not part of that.

rianquinn commented on bfvmm/include/hve/arch/intel_x64/xen/xen_op.h:
```diff
@@ -32,9 +32,13 @@
 #include <eapis/hve/arch/intel_x64/vmexit/wrmsr.h>
 #include <eapis/hve/arch/intel_x64/vmexit/rdmsr.h>
 #include <eapis/hve/arch/intel_x64/vmexit/io_instruction.h>
+#include <eapis/hve/arch/intel_x64/vmexit/ept_violation.h>
```
Same thing, can we get rid of the memory stuff now?

cjams replied:
No we're stuck with xAPIC if we want PVH

rianquinn commented on bfvmm/src/hve/arch/intel_x64/xen/xen_op.cpp:
```diff
@@ -357,6 +349,208 @@ xen_op_handler::exit_handler(
+}
+
+void
+xen_op_handler::xapic_handle_write_icr(uint64_t low)
```
I am assuming these can all be removed? Also, all of the APIC logic should be in the APIC class. I see the vAPIC a Hyperkernel thing, not a PV interface thing. So the RDMSR and WRMSR functions should all be in the APIC class and self contained.

cjams replied:
I don't agree. I think the object that represents a local APIC should have absolutely no notion of an exit or vtx or anything like that. I'm fine with making an apic handler class though

rianquinn replied:
Yeah, I don't really care about the details for now of how the lapic is represented internally. If you want to have another class that abstracts APIC stuff that's fine, but what I really care about is the lapic.cpp in the hyperkernel is (like the rest of the EAPIs code) is an lapic_handler that exposes an API that everything else uses. I don't want APIC logic (including exits) sprinkled throughout the code.

rianquinn commented on bfvmm/src/hve/arch/intel_x64/xen/xen_op.cpp:
```diff
@@ -544,6 +749,18 @@ xen_op_handler::cpuid_pass_through_handler(
+    using namespace ::intel_x64::cpuid::cache_parameters::eax;
+    bfignored(vcpu);
+
+    info.rax &= ~max_ids_logical::mask;
```
what about reserved fields?

rianquinn commented on bfvmm/src/hve/arch/intel_x64/xen/xen_op.cpp:
```diff
@@ -874,7 +1094,6 @@ xen_op_handler::XENVER_get_features_handler(
         info->submap |= (1 << XENFEAT_memory_op_vnode_supported);
         // info->submap |= (1 << XENFEAT_ARM_SMCCC_supported);
         info->submap |= (1 << XENFEAT_linux_rsdp_unrestricted);
-        info->submap |= (1 << XENFEAT_hvm_pirqs);
```
are you sure about this one? I would leave this on unless you have a specific reason to turn it off.

cjams replied:
Yep thats my bad

rianquinn commented on bfvmm/src/hve/arch/intel_x64/xen/xen_op.cpp:
```diff
@@ -913,6 +1132,7 @@ xen_op_handler::SCHEDOP_yield_handler(gsl::not_null<vcpu *> vcpu)
     try {
         m_sched_op->handle_yield(vcpu);
         vcpu->set_rax(SUCCESS);
+        m_vcpu->return_and_continue();
```
A yield would not be a return and continue. A yield should be a return and wait

cjams replied:
Yeah that stuff is old now; it will likely all be changed once we see sched_ops in their proper place

Commit:
Initial hkd commit

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Add hkd_* targets

- Add new targets similar to driver_build, driver_clean, etc.
  prefixed with hkd instead of driver

- Introduce script directory for cmake and utils scripts

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Add ioctl_mmap and ioctl_munmap

IOCTL_MMAP
----------
Allows userspace to allocate bytes from the kernel
with the promise that the memory will never be paged out of
memory and will already be paged in upon successful return.

That last bit is necessary so that the entries are valid when
the hypervisor walks the guest's page tables.

IOCTL_MUNMAP
------------
Unmaps the provided mmap

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Map in real-mode trampoline prior to launch

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Use mlockall+mmap for locked mappings

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Remove memory locking debugging code

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Implement VCPUOP_register_vcpu_time_memory_area

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Implement VCPUOP_register_runstate_memory_area

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Implement EVTCHNOP_expand_array

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Implement EVTCHNOP_set_priority

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Implement VCPUOP_set_singleshot_timer

Each time the guest makes this hypercall, we translate the
provided deadline to VMX-preemption timer ticks and arm the
timer.

The timer exit handler disarms the timer and performs an upcall
at the VIRQ_TIMER port that was bound previously in the
EVTCHNOP_bind_virq call.

Signed-off-by: Connor Davis <davisc@ainfosec.com>

rianquinn commented on bfvmm/src/hve/arch/intel_x64/xen/xen_op.cpp:
```diff
@@ -1147,6 +1170,39 @@ xen_op_handler::VCPUOP_stop_periodic_timer_handler(gsl::not_null<vcpu *> vcpu)
+        auto tsc_ticks = now - deadline;
+        auto pet_ticks = tsc_ticks >> m_pet_shift;
+
+        m_vcpu->set_vmx_preemption_timer(pet_ticks);
```
Don't forget that when you exit the VM for an external interrupt, you need to also wait in bfexec. This is a REALLY good sign BTW.

rianquinn replied:
Hopefully you see a yield soon

Commit:
Implement HYPERVISOR_vm_assist

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Start gnttab_op class

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Implement XENMEM_decrease_reservation

With this call, the guest passes a list of gfns to the hypervisor.
The hypervisor then unmaps and releases these gfns from the guests
physical address space.

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Implement XENMAPSPACE_grant_table

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Implement STORE_{PFN,EVTCHN} hvm params

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Use FIFO ABI for event channels

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
WIP: fixing singleshot timer bug

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Enable local xenstore

We abuse a bit in the shared info structure to signal
to the guest if it should setup a local xenstore. This
is a temporary hack in lieu of a long-term solution that
will likely take some coordination between linux/xen/bareflank
communities

Signed-off-by: Connor Davis <davisc@ainfosec.com>

Commit:
Add missing pieces for getting to userspace

- Add DSDT table
  Linux panics if there is no DSDT table. Provide
  an empty one and point to it from the FADT.

- Parse XEN_ELFNOTE_PHYS32_ENTRY
  The PVH entry point is not constant. Remove the
  hardcoded value in favor of reading the entry
  from the elfnote.

- Increase domu memory for initramfs unpacking
  The initramfs unpacking code fails with -ENOSPC
  if it runs out of memory. Increase the default
  RAM size for domu's in bfexec.

Signed-off-by: Connor Davis <davisc@ainfosec.com>
